ta/pkcs11/src/processing_symm.c

```diff
@@ -969,6 +1029,65 @@ enum pkcs11_rc step_symm_operation(struct pkcs11_session *session,
 			break;
 		}
 		break;
+	case PKCS11_CKM_AES_CBC_PAD:
+		if (step == PKCS11_FUNC_STEP_ONESHOT && !in_buf) {
+			EMSG("No input data");
+			return PKCS11_CKR_ARGUMENTS_BAD;
+		}
+
+		switch (function) {
+		case PKCS11_FUNCTION_ENCRYPT:
+			assert(proc->extra_ctx);
+
+			if (step == PKCS11_FUNC_STEP_ONESHOT) {
+				if (out_size < ((in_size & ~0xF) + TEE_AES_BLOCK_SIZE)) {
+					out_size = (in_size & ~0xF) + TEE_AES_BLOCK_SIZE;
+					rc = PKCS11_CKR_BUFFER_TOO_SMALL;
+					output_data = true;
+					break;
+				}
+				out_size_temp = out_size;
+				res = TEE_CipherUpdate(proc->tee_op_handle,
+						       in_buf, in_size,
+						       out_buf, &out_size);
+
+				output_data = true;
+				rc = tee2pkcs_error(res);
+				if (rc)
+					break;
+
+				/* Update text size still in the OS buffer. */
+				last_block->last_block_size = in_size & 0xF;
+				out_buf = (uint8_t *)out_buf + out_size;
+				out_size = out_size_temp - out_size;
+				out_size_temp -= out_size;
+			}
+
+			in_size_padded = TEE_AES_BLOCK_SIZE - last_block->last_block_size;
+
+			TEE_MemFill(last_block->pad_block, in_size_padded, in_size_padded);
+
+			res = TEE_CipherDoFinal(proc->tee_op_handle,
+						last_block->pad_block, in_size_padded,
+						out_buf, &out_size);
+			out_size += out_size_temp;
+
+			output_data = true;
+			rc = tee2pkcs_error(res);
+
+			break;
+		case PKCS11_FUNCTION_DECRYPT:
+			res = TEE_CipherDoFinal(proc->tee_op_handle,
+						in_buf, in_size,
+						out_buf, &out_size);
```

Comment:
My question may be stupid: is this implementation expected to successfully decrypt an AES CBC padded cipher text? I guess some post-processing is needed, maybe more.
If this piece of code is not expected to be reviewed, could you remove implementation of case`PKCS11_FUNCTION_DECRYPT` until you expect we review it.

Reply:
Sorry for the delay. I am having the decrypt under internal review right now. Is there a way to reopen this pull request?
The problem for decrypt I could not solve is without decryption it can only return the out_size including padding. Is it acceptable?

Reply:
For sure, decryption should return the right clear text size, excluding padded bytes artifact.
The issue I foresee is that caller is required to provide an output buffer that is large enough to hold those padded bytes, even if effective output buffer size will exclude these extra bytes. The weird thing is that a caller knowing the expected end size (without padding) and providing a buffer that excludes them will get a TEE_ERROR_SHORT_BUFFER he/she would not expect. This could be an compatibility issue with existing software packages that don't handle this.

A way out could be PKCS11 TA to always decrypt last block (at each step/update) into a private secure buffer and copy to effective output buffer when expected. That would add a bit of logic in the implementation but would behave better with any client implementation.

Another way out could be to manage that from cryptoki client library that would decrypt AES_CBC_PAD in temporary buffer, enlarged by 1 AES block.

Reply:
A temp buffer at client side might be the way to go. If my memory is not wrong, AES_CBC decryption normally needs cypher text of previous block. Only decrypt the last cypher block would not return correct plain text.

Reply:
Did some experiments. I do not think C_DecryptFinal can be used to remove padding. It does not take encrypt bytes. Meanwhile the encrypt text length of C_DecryptUpdate has to be multiple of 16, which means the decrypt buff length should be same as encrypt text length or else CKR_BUFFER_TOO_SMALL will return. So the decrypt buffer length of C_DecryptFinal can only be 0.